### DISCUSSION TITLE
RTSP bad image quality

### DISCUSSION THREAD

**Discussion by Xyprime** [roleoroleo/yi-hack-Allwinner-v2#790] [Q&A]
Hello,

I am currently facing the problem that my rtsp stream gives  bad image quality. I set up the camera(YI home 1080p) normaly like on the Website and used a python code(displayed at the end) to read the rtsp stream i use the high Res stream but still i get very bad image Qualtiy. Please give me some tips to enhance the image qualtiy. I use the Python code for face recognition. Python Code: 

import cv2
import face_recognition
import pickle
import time
import os
import numpy as np

def load_known_encodings():
    try:
        with open("known_encodings.pkl", "rb") as file:
            return pickle.load(file)
    except FileNotFoundError:
        return {}

def load_known_people_data():
    try:
        with open("known_people_data.pkl", "rb") as file:
            return pickle.load(file)
    except FileNotFoundError:
        return {}

def draw_face_info(frame, name, confidence, extra_info, top, right, bottom, left):
    outline_color = (0, 255, 0) if name != "Unknown" else (0, 0, 255)
    fill_opacity = 0.3 if name == "Unknown" else 0

    cv2.rectangle(frame, (left, top), (right, bottom), outline_color, cv2.FILLED)
    cv2.rectangle(frame, (left, top), (right, bottom), (0, 0, 0), 2)

    font = cv2.FONT_HERSHEY_DUPLEX
    font_size = 0.8
    font_thickness = 2

    cv2.putText(frame, f"{name} ({confidence * 100:.1f}%)", (left + 6, bottom - 30), font, font_size, (0, 0, 0),
                font_thickness)

    line_height = 20

    if "age" in extra_info:
        cv2.putText(frame, f"Age: {extra_info['age']}", (left + 6, top + 20), font, font_size, (0, 0, 0),
                    font_thickness)

    if "gender" in extra_info:
        cv2.putText(frame, f"Gender: {extra_info['gender']}", (left + 6, top + 21 + line_height), font, font_size,
                    (0, 0, 0), font_thickness)

    if "race" in extra_info:
        cv2.putText(frame, f"Race: {extra_info['race']}", (left + 6, top + 22 + 2 * line_height), font, font_size,
                    (0, 0, 0), font_thickness)

    if "school" in extra_info:
        cv2.putText(frame, f"School: {extra_info['school']}", (left + 6, top + 22 + 3 * line_height), font, font_size,
                    (0, 0, 0), font_thickness)

def recognize_faces(frame, known_encodings, known_people_data,i):
    small_frame = cv2.resize(frame, (0, 0), fx=0.25, fy=0.25)
    face_locations = face_recognition.face_locations(small_frame)
    ret, frame1 = i.read()
    face_encodings = face_recognition.face_encodings(small_frame, face_locations)

    matches = {name: face_recognition.compare_faces(encodings, face_encoding, tolerance=0.6) for name, encodings in
               known_encodings.items() for face_encoding in face_encodings}

    for face_encoding, (top, right, bottom, left) in zip(face_encodings, face_locations):
        name = "Unknown"
        confidence = 0
        extra_info = {}

        for person, match_list in matches.items():
            if any(match_list):
                confidence = match_list.count(True) / len(match_list)
                if confidence >= 0.95:
                    name = person
                    extra_info = known_people_data.get(name, {})
                    break

        draw_face_info(frame, name, confidence, extra_info, top * 4, right * 4, bottom * 4, left * 4)

def main():
    known_encodings = load_known_encodings()
    known_people_data = load_known_people_data()
    **rtsp_url = 'rtsp://192.168.178.168/ch0_0.h264'**

    **video_capture = cv2.VideoCapture(rtsp_url)
    start_time = time.time()****
    frame_count = 0

    while True:
        ret, frame = video_capture.read()

        recognize_faces(frame, known_encodings, known_people_data,video_capture)

        frame_count += 1
        elapsed_time = time.time() - start_time
        fps = frame_count / elapsed_time

        cv2.putText(frame, f"FPS: {fps:.2f}", (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1)

        cv2.imshow('Video', frame)

        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

    video_capture.release()
    cv2.destroyAllWindows()

if __name__ == "__main__":
    main()

**Comment by roleoroleo** [comment]
How is the quality if you use VLC?

**Reply by Xyprime** [reply]
it is the same on vlc and python

**Reply by Xyprime** [reply]
but the time diffrence between vlc and python is bigger python is like 10 seconds and vlc 2 seconds

**Comment by roleoroleo** [comment]
I recently made a lot of fixes in the RTSP server.
Tell me your model suffix, I will send a new beta release.

**Reply by Xyprime** [reply]
My Model is yi home 1080p

**Reply by Xyprime** [reply]
But I dont think it is an problem at your end neither at my end because the bad image qualtiy is the same even in the app but thanks for the help

**Reply by roleoroleo** [reply]
Ah ok. It's not a problem of frame corruption.

**Comment by nemonoid** [comment]
> But I dont think it is an problem at your end neither at my end because the bad image qualtiy is the same even in the app but thanks for the help

Yes, the image quality on these cameras is not great, it's 1080p but the bitrate is quite low.  
That said it can be OK, depending on the usercase..

**Comment by roleoroleo** [comment]
What your model suffix?
y21ga, h52ga, ...
